framework/quota/src/org/apache/cloudstack/quota/dao/QuotaTariffDaoImpl.java

```diff
@@ -0,0 +1,137 @@
+//KIND, either express or implied.  See the License for the
+//specific language governing permissions and limitations
+//under the License.
+package org.apache.cloudstack.quota.dao;
+
+import com.cloud.utils.db.Filter;
+import com.cloud.utils.db.GenericDaoBase;
+import com.cloud.utils.db.SearchBuilder;
+import com.cloud.utils.db.SearchCriteria;
+import com.cloud.utils.db.TransactionLegacy;
+
+import com.cloud.utils.exception.CloudRuntimeException;
+import org.apache.cloudstack.quota.constant.QuotaTypes;
+import org.apache.cloudstack.quota.vo.QuotaTariffVO;
+import org.apache.log4j.Logger;
+import org.springframework.stereotype.Component;
+
+import javax.ejb.Local;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+@Component
+@Local(value = { QuotaTariffDao.class })
+public class QuotaTariffDaoImpl extends GenericDaoBase<QuotaTariffVO, Long> implements QuotaTariffDao {
+    private static final Logger s_logger = Logger.getLogger(QuotaTariffDaoImpl.class.getName());
+
+    private final SearchBuilder<QuotaTariffVO> searchUsageType;
+    private final SearchBuilder<QuotaTariffVO> listAllIncludedUsageType;
+
+    public QuotaTariffDaoImpl() {
+        super();
+        searchUsageType = createSearchBuilder();
+        searchUsageType.and("usage_type", searchUsageType.entity().getUsageType(), SearchCriteria.Op.EQ);
+        searchUsageType.done();
+
+        listAllIncludedUsageType = createSearchBuilder();
+        listAllIncludedUsageType.and("onorbefore", listAllIncludedUsageType.entity().getEffectiveOn(), SearchCriteria.Op.LTEQ);
+        listAllIncludedUsageType.and("quotatype", listAllIncludedUsageType.entity().getUsageType(), SearchCriteria.Op.EQ);
+        listAllIncludedUsageType.done();
+    }
+
+    @Override
+    public QuotaTariffVO findTariffPlanByUsageType(final int quotaType, final Date effectiveDate) {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        List<QuotaTariffVO> result = new ArrayList<>();
+        try (TransactionLegacy txn = TransactionLegacy.open(TransactionLegacy.USAGE_DB)) {
+            final Filter filter = new Filter(QuotaTariffVO.class, "effectiveOn", false, 0L, 1L);
+            final SearchCriteria<QuotaTariffVO> sc = listAllIncludedUsageType.create();
+            sc.setParameters("onorbefore", effectiveDate);
+            sc.setParameters("quotatype", quotaType);
+            result = search(sc, filter);
+        } catch (Exception e) {
+            throw new CloudRuntimeException("Unable to find tariff plan by usage type");
+        } finally {
+            TransactionLegacy.open(opendb).close();
+        }
+        if (result.size() > 0) {
+            if (s_logger.isDebugEnabled()){
+                s_logger.debug("QuotaTariffDaoImpl::findTariffPlanByUsageType: " + effectiveDate + "quota type " + quotaType  + " val=" + result.get(0).getCurrencyValue());
+            }
+            return result.get(0);
```

Comment:
What happens if/when the result is greater than zero?  If this condition is unexpected, an ``IllegalStateException`` should be thrown when it occurs.